### PR TITLE
Detection Module (RetinaNet)

### LINKED ISSUE #3571: Box utils components
**Is your feature request related to a problem? Please describe.**
The util components used for detection box are not in MONAI

**Describe the solution you'd like**
Add box util components

**Describe alternatives you've considered**
N/A

**Additional context**

### LINKED ISSUE #3572: Box transforms
**Is your feature request related to a problem? Please describe.**
Box pre- and post-processing transform functions are not in MONAI

**Describe the solution you'd like**
Add box transform functions

**Describe alternatives you've considered**
N/A

**Additional context**
1. convert input box with original mode to the box with standard mode
2. convert output box with standard mode to the box with original mode
3. Flip boxes with image as a transform

### LINKED ISSUE #3573: Anchor generator
**Is your feature request related to a problem? Please describe.**
Anchor generator for box detection is not in MONAI

**Describe the solution you'd like**
Add anchor generator

**Describe alternatives you've considered**
N/A

**Additional context**

### LINKED ISSUE #3576: Box selector
**Is your feature request related to a problem? Please describe.**
A box selector for detection postprocessing is required

**Describe the solution you'd like**
Add box selector

**Describe alternatives you've considered**
N/A

**Additional context**
1. Top box selector
2. Non-maximum suppression selector

### LINKED ISSUE #3748: Box evaluation metrics
**Is your feature request related to a problem? Please describe.**
No box evaluator

**Describe the solution you'd like**
Add box evaluation metrics and evaluator

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### LINKED ISSUE #3777: Sliding window inferer for detection
**Is your feature request related to a problem? Please describe.**
Current sliding window inferer does not work for detection

**Describe the solution you'd like**
Add a sliding window inferer that works for multiple outputs and outputs that have different size compared with input

**Describe alternatives you've considered**
N/A

**Additional context**
N/A

### PR DESCRIPTION
**Pull request by Can-Zhao**
Fixes #3571, Fixes #3572, Fixes #3576, Fixes #3573, Fixes #3748, Fixes #3777

### Description
Add box util components and corresponding unit test;
Add Box transforms that are necessary for basic run and corresponding unit test;

Below are 3D version of torchvision implementation:
Add anchor generator;
Add retinanet;
Add loss computation;
Add post-processing;
Add sliding window inferer for multi-output preditor;
Enable sliding window inference for detection;

Below are from nnDetection:
Add box evaluation metrics

(need further polishing. Current version could provide reasonable result on synthetic 3D data.)

### Status
**Work in progress**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [ ] Breaking change (fix or new feature that would cause existing functionality to change).
- [x] New tests added to cover the changes.
- [ ] Integration tests passed locally by running `./runtests.sh -f -u --net --coverage`.
- [ ] Quick tests passed locally by running `./runtests.sh --quick --unittests  --disttests`.
- [ ] In-line docstrings updated.
- [ ] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

File: monai/data/box_utils.py [hunk 1]
```diff
@@ -0,0 +1,385 @@
...
+    mode: choose from SUPPORT_MODE. If mode is not given, these funcs will assume mode is STANDARD_MODE
+    image_size: Length of 2 or 3. Data format is list, or np.ndarray, or tensor of int
+"""
+
+
+def check_support_mode(mode):
+    """
+    Check if the mode is supported
+    """
+    if mode not in SUPPORT_MODE:
```
Comment (wyli) [inline monai/data/box_utils.py:40]: there's a `look_up_option ` utility, please make use of it... https://github.com/Project-MONAI/MONAI/blob/3c6ee694d5e7fd6d9da2c65ed1b83030488da064/monai/utils/module.py#L47-L74

there are also some other utilities, might be useful https://github.com/Project-MONAI/MONAI/blob/dev/monai/utils/misc.py#L29-L48

File: tests/test_box_utils.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
+# Copyright 2020 - 2021 MONAI Consortium
```
Comment (wyli) [inline tests/test_box_utils.py:1]: now we use `# Copyright (c) MONAI Consortium` instead of the `2020 - 2021`

https://github.com/Project-MONAI/MONAI/blob/dev/CONTRIBUTING.md#preparing-pull-requests

File: monai/data/box_utils.py [hunk 3]
```diff
@@ -0,0 +1,368 @@
...
+    """
+    convert a torch.Tensor, or np array input to list
+    Args:
+        in_sequence:
+    Returns: in_sequence_list
+
+    """
+    in_sequence_list = deepcopy(in_sequence)
+    if torch.is_tensor(in_sequence):
+        in_sequence_list.cpu().detach().numpy().tolist()
```
Comment (mibaumgartner) [inline monai/data/box_utils.py:48]: Nit: `.detach().cpu()` might be lightly better

File: monai/data/box_utils.py [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+    mode: choose from SUPPORT_MODE. If mode is not given, these funcs will assume mode is STANDARD_MODE
+    image_size: Length of 2 or 3. Data format is list, or np.ndarray, or tensor of int
+"""
+
+
+def convert_to_list(in_sequence: Union[Sequence, torch.Tensor, np.ndarray]) -> list:
+    """
+    convert a torch.Tensor, or np array input to list
+    Args:
+        in_sequence:
```
Comment (mibaumgartner) [inline monai/data/box_utils.py:42]: Nit: Doc missing: what is the expected input format of the sequence?

File: monai/data/box_utils.py [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+        mode2 = get_standard_mode(int(bbox2.shape[1] / 2))
+    mode1 = look_up_option(mode1,supported=STANDARD_MODE)
+    mode2 = look_up_option(mode2,supported=STANDARD_MODE)
+    spatial_dims = get_dimension(bbox=bbox1, mode=mode1)
+
+    # we do computation on cpu
+    device = bbox1.device
+
+    # compute area for the bbox
+    area1 = box_area(bbox=bbox1, mode=mode1).cpu()  # Nx1
```
Comment (mibaumgartner) [inline monai/data/box_utils.py:350]: Why is the area moved to CPU here?

File: monai/data/box_utils.py [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+    area2 = box_area(bbox=bbox2, mode=mode2).cpu()  # Mx1
+
+    # get the left top and right bottom points for the NxM combinations
+    lt = torch.max(bbox1[:, None, ::2], bbox2[:, ::2])  # [N,M,spatial_dims] left top
+    rb = torch.min(bbox1_corner[:, None, 1::2], bbox2_corner[:, 1::2])  # [N,M,spatial_dims] right bottom
```
Comment (mibaumgartner) [inline monai/data/box_utils.py:355]: I can't find bbox1_corner | bbox2_corne, where is this defined?

File: monai/data/box_utils.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+    # compute size for the intersection region for the NxM combinations
+    wh = (rb - lt + TO_REMOVE).clamp(min=0)  # [N,M,spatial_dims]
+    inter = wh[:, :, 0]  # [N,M]
+    for axis in range(1, spatial_dims):
+        inter = inter * wh[:, :, axis]
```
Comment (mibaumgartner) [inline monai/data/box_utils.py:360]: Note/Warning: Might overflow in 3D with Mixed Precision Training

File: monai/transforms/box/array.py [hunk 8]
```diff
@@ -0,0 +1,158 @@
...
+        else:
+            bbox_tensor = bbox
+
+        if self.spatial_axis is None:
+            self.spatial_axis = list(range(0, len(image_size)))
+        if isinstance(self.spatial_axis, int):
+            self.spatial_axis = [self.spatial_axis]
+
+        # clip box to the image and (optional) remove empty box
+        bbox_clip = box_utils.box_clip_to_image(bbox_tensor, image_size, self.mode, self.remove_empty)
```
Comment (mibaumgartner) [inline monai/transforms/box/array.py:148]: Nit: might want to add the clipping to the docs of this function

**Review by mibaumgartner — COMMENTED** [review]
I left some comments, the PR looks great so far!
Adding some tests with `TO_REMOVE =1` might be a nice addition as well.

**Comment by Can-Zhao** [comment]
> I left some comments, the PR looks great so far! Adding some tests with `TO_REMOVE =1` might be a nice addition as well.

Thank you!
I added comment that only TO_REMOVE =0 has been tested for now. We may got time to test TO_REMOVE =1 later.

File: monai/data/box_utils.py [hunk 9]
```diff
@@ -0,0 +1,383 @@
...
+        mode = get_standard_mode(int(bbox.shape[1] / 2))
+    mode = look_up_option(mode, supported=STANDARD_MODE)
+    spatial_dims = get_dimension(bbox=bbox, mode=mode)
+
+    area = bbox[:, 1] - bbox[:, 0] + TO_REMOVE
+    for axis in range(1, spatial_dims):
+        area = area * (bbox[:, 2 * axis + 1] - bbox[:, 2 * axis] + TO_REMOVE)
+
+    if torch.isnan(area).any() or torch.isinf(area).any():
+        if area.dtype in {torch.bfloat16, torch.float16}:
```
Comment (mibaumgartner) [inline monai/data/box_utils.py:274]: 1) I'm not sure about `bfloat16` here since it should support the same value range as float32 (just with lower "resolution")
2) I think there is no 100% clean solution for this problem but it is probably a good idea to deactivate mixed precision for these computations at some point. This could either be left to the object calling this function or this function. Usually, it is not desired to put the whole network in full precision since that would consume a lot of Memory. Both places enforce the usage of a specific Mixed Precision training framework though.

Deactivation code for Native Pytorch: `@autocast(enabled=False)` (imported from `from torch.cuda.amp import autocast`) followed by a manual cast to float32. <- This solution only works for Native PyTorch Mixed Precision (which is probably the default for most people by now) but (I think) it won't work for APEX MP.

File: monai/data/box_utils.py [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+    inter = wh[:, :, 0]  # [N,M]
+    for axis in range(1, spatial_dims):
+        inter = inter * wh[:, :, axis]
+
+    # compute IoU
+    iou = inter / (area1[:, None] + area2 - inter + torch.finfo(area1.dtype).eps)  # [N,M,spatial_dims]
+
+    if torch.isnan(iou).any() or torch.isinf(iou).any():
+        if iou.dtype in {torch.bfloat16, torch.float16}:
+            ValueError(f"Box IoU is aN or Inf. torch.float16 is used. Please change to torch.float32 and test it again.")
```
Comment (mibaumgartner) [inline monai/data/box_utils.py:376]: Same as above and a small typo `NaN`.

File: tests/test_box_utils.py [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+        [
+            {"bbox": p(bbox), "image_size": image_size, "mode": "xxyyzz"},
+            "xyzwhd",
+            [[0, 0, 0, 0, 0, 0], [0, 0, 2, 1, 2, 1], [0, 1, 2, 1, 1, 1]],
+            [0, 2, 1],
+        ]
+    )
+
+
+class TestCreateBoxList(unittest.TestCase):
```
Comment (mibaumgartner) [inline tests/test_box_utils.py:99]: Nit: might be a good idea to write separate unit tests for the individual components to isolate errors from each other (this is a pure design decision though).

File: tests/test_box_utils.py [hunk 12]
```diff
@@ -0,0 +1,124 @@
...
+
+        result_standard = box_convert_standard_mode(bbox=bbox1, mode=mode1)
+        expected_box_standard = box_convert_standard_mode(bbox=expected_box, mode=mode2)
+        self.assertEqual(result_standard.cpu().numpy().tolist(), expected_box_standard.cpu().numpy().tolist())
+
+        self.assertEqual(box_area(result_standard).tolist(), expected_area)
+
+        result_standard_clip = box_clip_to_image(result_standard, image_size, mode=None, remove_empty=True)
+        self.assertEqual(box_area(result_standard_clip).tolist(), list(filter(lambda num: num > 0, expected_area)))
+
```
Comment (mibaumgartner) [inline tests/test_box_utils.py:121]: `box_iou` test is missing

File: monai/metrics/detection/coco_evaluator.py [hunk 13]
```diff
@@ -0,0 +1,435 @@
...
+        for max_det_idx, max_det in enumerate(self.max_detections):  # mAR
+            key = f"mAR_IoU_{self.iou_range[0]:.2f}_{self.iou_range[1]:.2f}_{self.iou_range[2]:.2f}_MaxDet_{max_det}"
+            results[key] = self.select_ar(dataset_statistics, max_det_idx=max_det_idx)
+
+            if self.per_class:
+                for cls_idx, cls_str in enumerate(self.classes):  # per class results
+                    key = (f"{cls_str}_"
+                           f"mAR_IoU_{self.iou_range[0]:.2f}_{self.iou_range[1]:.2f}_{self.iou_range[2]:.2f}_"
+                           f"MaxDet_{max_det}")
+                    results[key] = self.select_ar(dataset_statistics,
```
Comment (mibaumgartner) [inline monai/metrics/detection/coco_evaluator.py:214]: Same as above

**Review by mibaumgartner — COMMENTED** [review]
Found an indexing bug in mAR where all IoU values are selected for averaging instead of just the indicated ones.
